### DISCUSSION TITLE
Option to bring back the Assets sidebar folders view

### DISCUSSION THREAD

**Discussion by elivz** [craftcms/cms#13535] [Ideas] [answered]
We recently upgraded a client's site and are getting some very negative feedback regarding the UX changes to the Assets index. They vastly preferred the old layout – being able to see the full sub-folder tree all at once rather than diving in and out to see the different levels. They are wondering if there is any possibility of an option to turn back on the folder display under each of the volumes in the sidebar.

> I **hate** this new way (and it’s not just a matter of getting used to it). We had everything organized in a way that made sense and made it easier to access/upload whatever we needed in an organized way (by categories, collections, products, resources for retailers etc. through the convenient pulldown). This has me clicking through a ton of folders trying to find the right subfolder etc, versus the old way that just allowed me to click the destination, and easily create/see/access subfolders in an intuitive way.

**Comment by brandonkelly** [comment, marked as answer]
I don’t think we should have an officially supported way of bringing subfolders back into the source list, given the [reasons](https://craftcms.com/blog/rethinking-volume-subfolders) we changed it in the first place. But modules/plugins are free to modify element sources via `EVENT_REGISTER_SOURCES`.

Here’s how you could do it in Craft 4.5 (pending [this change](https://github.com/craftcms/cms/commit/cbe96b2209ce294ec84a23ce45c0871efbac1cfd)):

```php
use craft\base\Event;
use craft\elements\Asset;
use craft\events\RegisterElementSourcesEvent;
use craft\models\VolumeFolder;

// ...

private function attachEventHandlers(): void
{
    Event::on(Asset::class, Asset::EVENT_REGISTER_SOURCES, function(RegisterElementSourcesEvent $event) {
        $assetsService = Craft::$app->getAssets();

        foreach ($event->sources as &$source) {
            if (isset($source['data']['folder-id'])) {
                $folder = $assetsService->getFolderById($source['data']['folder-id']);
                if ($folder && !$folder->parentId) {
                    $subfolders = $assetsService->getAllDescendantFolders($folder, withParent: false, asTree: true);
                    $source['nested'] = $this->folders2sources($subfolders, $source);
                }
            }
        }
    });
}

private function folders2sources(array $folders, array $rootSource): array
{
    return array_map(fn(VolumeFolder $folder) => [
        'key' => "folder:$folder->uid",
        'label' => $folder->name,
        'hasThumbs' => true,
        'criteria' => ['folderId' => $folder->id],
        'defaultSort' => ['dateCreated', 'desc'],
        'data' => [
            'folder-id' => $folder->id,
            'can-upload' => $rootSource['data']['can-upload'] ?? false,
            'can-move-to' => $rootSource['data']['can-move-to'] ?? false,
            'can-move-peer-files-to' => $rootSource['data']['can-move-peer-files-to'] ?? false,
            'fs-type' => $rootSource['fs-type'] ?? null,
        ],
        'nested' => $this->folders2sources($folder->getChildren(), $rootSource),
    ], $folders);
}
```

Functionality for nested folders selected via the source list would be limited – you wouldn’t be able to rename/delete them or create new subfolders. Bringing those actions back would be more involved.

**Reply by elivz** [reply]
Amazing, thanks Brandon! This site is actually still on 3.8, but I imagine I can figure it out with this example as a starting point.

**Reply by brandonkelly** [reply]
Just backported the change to the 3.9 branch as well (9a4c92c2e2ab1848f67951e875954bbed71c74d4). You can get on that by changing your `craftcms/cms` requirement to `3.9.x-dev as 3.9.0-alpha`.

**Reply by elivz** [reply]
You're the best, Brandon!

**Reply by brandonkelly** [reply]
3.9.0 and 4.5.0 are out now with that change 🎉

**Reply by elivz** [reply]
FYI for anyone else implementing this solution: I ended up having to memoize the results of those `folders2sources` calls to avoid a timeout on entry pages containing a lot of Assets fields (this code gets run separately for each Assets field present on the page, so on a site with a lot of subfolders there can be a _lot_ of duplicate processing happening).
